docs/en/engines/table-engines/mergetree-family/replacingmergetree.md

```diff
@@ -100,8 +100,43 @@ SELECT * FROM mySecondReplacingMT FINAL;
 The row is deleted when `OPTIMIZE ... FINAL CLEANUP` or `OPTIMIZE ... FINAL` is used, or if the engine setting `clean_deleted_rows` has been set to `Always`.
 
 No matter the operation on the data, the version must be increased. If two inserted rows have the same version number, the last inserted row is the one kept.
+
+Always execute `OPTIMIZE ... FINAL CLEANUP` and `OPTIMIZE ... FINAL` to delete rows with `is_deleted=1` defined, especially when you wish to insert a new row with an older version. Otherwise, the new row with an older version will be replaced and not be persisted.
```

Comment:
A very strange sentence. I don't understand it.
Always??
`is_deleted` are removed during the merge **IF** a new data is inserted. No need to run optimize at all.
And you should never insert older versions if you want to store them, because ReplacingMergeTree stores latest versions.

Reply:
Yes, this sentence is confusing. You don't need to run OPTIMIZE if `clean_deleted_rows` is set to `Always` (see the sentence a few lines above)

Also, I don't understand what does this mean:
> Otherwise, the new row with an older version will be replaced and not be persisted.

cc: @DerekChia

Reply:
This is what I meant, and to provide some warning on this behavior:

In both setup, I use the same table DDL:
```
CREATE OR REPLACE TABLE myThirdReplacingMT
(
    `key` Int64,
    `someCol` String,
    `eventTime` DateTime,
    `is_deleted` UInt8
)
ENGINE = ReplacingMergeTree(eventTime, is_deleted)
ORDER BY key;
```

Without running `OPTIMIZE TABLE myThirdReplacingMT FINAL CLEANUP` before the 3rd insert, we will not see any rows. i.e. 3rd is lost
```
INSERT INTO myThirdReplacingMT Values (1, 'first', '2020-01-01 01:01:01', 0);
INSERT INTO myThirdReplacingMT Values (1, 'first', '2020-01-01 01:01:01', 1); 

INSERT INTO myThirdReplacingMT Values (1, 'first', '2020-01-01 00:00:00', 0);
select * from myThirdReplacingMT final; 
0 rows in set. Elapsed: 0.002 sec.
```

With `OPTIMIZE TABLE myThirdReplacingMT FINAL CLEANUP`, we will see the 3rd row.
```
INSERT INTO myThirdReplacingMT Values (1, 'first', '2020-01-01 01:01:01', 0);
INSERT INTO myThirdReplacingMT Values (1, 'first', '2020-01-01 01:01:01', 1); 

OPTIMIZE TABLE myThirdReplacingMT FINAL CLEANUP; 
INSERT INTO myThirdReplacingMT Values (1, 'first', '2020-01-01 00:00:00', 0);
select * from myThirdReplacingMT final; 

┌─key─┬─someCol─┬───────────eventTime─┬─is_deleted─┐
│   1 │ first   │ 2020-01-01 00:00:00 │          0 │
└─────┴─────────┴─────────────────────┴────────────┘

1 row in set. Elapsed: 0.003 sec.
```

After inserting the 2nd row, many would have thought that eventTime=1 has been deleted (but it isn't) and will proceed to insert the 3rd row only to realize it is not inserted. So I aim to provide some warning there. 

Maybe I'm doing something wrong. Thoughts, please? @tavplubix @den-crane

Reply:
But this behavior is already documented: https://github.com/ClickHouse/ClickHouse/blob/eeacdbae9c15554ca5c05f7e80165d4559ab9abd/docs/en/engines/table-engines/mergetree-family/replacingmergetree.md?plain=1#L91-L102

By default `clean_deleted_rows` is set to `Never` to avoid this issue. The third row should not resurrect. 

> After inserting the 2nd row, many would have thought that eventTime=1 has been deleted (but it isn't) and will proceed to insert the 3rd row only to realize it is not inserted.

The expected behavior is that the third row should not be shown on `SELECT FINAL`. Because its "version" is old, and ReplacingMergeTree only keeps rows with the newest version.

Reply:
Thanks, I guess this is the line that provides the warning. 
> No matter the operation on the data, the version must be increased. If two inserted rows have the same version number, the last inserted row is the one kept.